---
title: Research
layout: page
permalink: /research/
# image: assets/images/research.png
---

In broad terms, my interests lie in understanding the molecular mechanisms underlying the production of phenotypic variation in complex traits. During my Ph.D. with [Todd Castoe](https://www.castoelaboratory.org/), my focus was on investigating the production of phenotypic variation in the expression of toxins in snake venom. Snake venoms are a notable evolutionary system because of the high degree of variation in venom composition, even between individuals of the same species. Studying this topic means combining inferences derived from several different functional genomic techniques, such as RNA-seq, ATAC-seq, single-cell methods, whole genome sequencing data and more. Below, I go into more detail on specific research themes and work I've been involved with.

<br>

## Divergent gene regulation and phenotypic variation in non-model systems

![left-aligned-image]({{ site.url }}{{ site.baseurl }}/assets/images/research1.png){: .align-left}
A unifying theme across many fields of biology is understanding how changes in genotype cause changes in phenotype. My doctoral work was centered on this topic. My first contribution was to a study that identified many of the genomic sequences and transcription factors that control the production of rattlesnake venom [(Perry et al. 2022)](https://www.genome.org/cgi/doi/10.1101/gr.276251.121). This directly led to supplemental work characterizing the genomic structure of major venom component in rattlesnakes [(Gopalan et al., 2022)](https://doi.org/10.1016/j.toxicon.2022.06.014) and understanding how regulatory network variation across a phylogenetic sampling drove divergent venom expression profiles [(Gopalan et al. 2024)](https://doi.org/10.1093/gbe/evae110).

<br>

## Using single-cell methods to detect variation-producing mechanisms

Single-cell sequencing methods are a relatively new technique that we've used to identify molecular pathways that produce tissue-level expression heterogeneity [(Westfall et al., 2022)](https://doi.org/10.1093/gbe/evad109). 

As a hypothesis testing platform, the predictive power of single-cell sequencing is affected by the number of cells sampled and the types of data that can be measured at once. Future directions for this work are to leverage the combined measurement of gene expression and chromatin state to precisely measure how changing genome accessiblity affects gene expression. This type of "multi-omic" assay provides an incredible increase in power to detect cis-regulatory elements compared to other tissue-based methods that will provide very detailed maps of regulatory interactions in the genome.
